subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
             return false;
         }
     }
+
+    private void detectMissingDependencies(LocalTaskNode node) {
+        RelatedLocations consumedDirectories = consumedAndProducedLocations.getConsumedDirectories();
+        for (String outputPath : node.getMutationInfo().outputPaths) {
+            consumedDirectories.getNodesRelatedTo(outputPath).stream()
+                .filter(consumerNode -> missesDependency(node, consumerNode))
+                .forEach(consumerWithoutDependency -> emitMissingDependencyDeprecationWarning(node, consumerWithoutDependency));
+        }
+        Set<String> consumedLocations = new LinkedHashSet<>();
+        node.getTaskProperties().getInputFileProperties()
+            .forEach(spec -> spec.getPropertyFiles().visitStructure(new FileCollectionStructureVisitor() {
+                @Override
+                public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
+                    contents.forEach(location -> consumedLocations.add(location.getAbsolutePath()));
+                }
+
+                @Override
+                public void visitGenericFileTree(FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    fileTree.forEach(location -> consumedLocations.add(location.getAbsolutePath()));
+                }
+
+                @Override
+                public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
+                    consumedLocations.add(root.getAbsolutePath());
+                }
+
+                @Override
+                public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    consumedLocations.add(file.getAbsolutePath());
+                }
+            }));
+        consumedDirectories.recordRelatedToNode(node, consumedLocations);
+        for (String consumedLocation : consumedLocations) {
+            consumedAndProducedLocations.getProducedDirectories().getNodesRelatedTo(consumedLocation).stream()
+                .filter(producerNode -> missesDependency(producerNode, node))
+                .forEach(producerWithoutDependency -> emitMissingDependencyDeprecationWarning(producerWithoutDependency, node));
+        }
```

Comment:
I am currently collecting the elements first and then iterate over them afterwards, so I only check each location once. Checking each location directly when I collect it caused a performance regression in `ArchiveTreePerformanceTest`, since we add the same location 1000 times as an input.